# Pyspark Tutorial

This is a DataCamp tutorial on PySpark

https://www.datacamp.com/tutorial/pyspark-tutorial-getting-started-with-pyspark

# Installing Libraries

Ensure pyspark is installed in the environment 

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql import functions as F # as F so some of the functions won't get confused with inbuilt functions
from pyspark.sql.types import *

# Pandas
import pandas as pd

# Creating a SparkSession

A SparkSession is an entry point into all functionality in Spark, and is required if you want to build a dataframe in PySpark. Run the following lines of code to initialize a SparkSession:

In [2]:
from pyspark.sql import SparkSession 

spark = (
    SparkSession.builder
    .appName("Datacamp Pyspark Tutorial")
    .config("spark.memory.offHeap.enabled", "true")
    .config("spark.memory.offHeap.size", "10g")
    .getOrCreate()
)

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/10/21 13:20:37 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


Using the codes above, we built a spark session and set a name for the application. Then, the data was cached in off-heap memory to avoid storing it directly on disk, and the amount of memmory was manually specified.

# Creating the DataFrame

Data is mamually downloaded from the following link: https://archive.ics.uci.edu/dataset/352/online+retail

In [3]:
import pandas as pd
pandas_df = pd.read_excel("datacamp_ecommerse.xlsx") # There's no modules for reading xlsx files in Spark so you need to use Pandas first
df = spark.createDataFrame(pandas_df) # Convert pandas dataframe into Spark dataframe

In [4]:
df.show(5, 0)

25/10/21 13:21:11 WARN TaskSetManager: Stage 0 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------+---------+-----------------------------------+--------+-------------------+---------+----------+--------------+
|InvoiceNo|StockCode|Description                        |Quantity|InvoiceDate        |UnitPrice|CustomerID|Country       |
+---------+---------+-----------------------------------+--------+-------------------+---------+----------+--------------+
|536365   |85123A   |WHITE HANGING HEART T-LIGHT HOLDER |6       |2010-12-01 08:26:00|2.55     |17850.0   |United Kingdom|
|536365   |71053    |WHITE METAL LANTERN                |6       |2010-12-01 08:26:00|3.39     |17850.0   |United Kingdom|
|536365   |84406B   |CREAM CUPID HEARTS COAT HANGER     |8       |2010-12-01 08:26:00|2.75     |17850.0   |United Kingdom|
|536365   |84029G   |KNITTED UNION FLAG HOT WATER BOTTLE|6       |2010-12-01 08:26:00|3.39     |17850.0   |United Kingdom|
|536365   |84029E   |RED WOOLLY HOTTIE WHITE HEART.     |6       |2010-12-01 08:26:00|3.39     |17850.0   |United Kingdom|
+---------+-----

Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [5]:
df.printSchema()

root
 |-- InvoiceNo: string (nullable = true)
 |-- StockCode: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Quantity: long (nullable = true)
 |-- InvoiceDate: timestamp (nullable = true)
 |-- UnitPrice: double (nullable = true)
 |-- CustomerID: double (nullable = true)
 |-- Country: string (nullable = true)



The DataFrame consists of 8 variables:

1. InvoiceNo: The unique identifier of each customer invoice
2. StockCode: The unique identifier of each item in stock
3. Description: The item purchased by the customer
4. Quantity: The number of each item purchased by a customer in a single invoice
5. InvoiceDate: The purchase date
6. UnitPrice: Price of one unit of each item
7. CustomerID: Unique identifier assigned to each user
8. Country: The country from where the purchase was made

# Data Exploration

In [6]:
df.count()

25/10/21 13:21:11 WARN TaskSetManager: Stage 1 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


541909

There are 541,909 rows in the dataframe

## Missing Values

In [7]:
df.columns

['InvoiceNo',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'UnitPrice',
 'CustomerID',
 'Country']

In [8]:
# String Columns
str_col = ['InvoiceNo','StockCode','Description','InvoiceDate', 'Country'] # 5 columns
df.select([
    F.count(
        F.when(F.col(c).isNull(), c)
    ).alias(c) for c in str_col
]).show()

25/10/21 13:21:12 WARN TaskSetManager: Stage 4 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------+---------+-----------+-----------+-------+
|InvoiceNo|StockCode|Description|InvoiceDate|Country|
+---------+---------+-----------+-----------+-------+
|        0|        0|          0|          0|      0|
+---------+---------+-----------+-----------+-------+



In [9]:
# Numerical Columns
str_col = ['Quantity','UnitPrice','CustomerID'] # 3 columns
df.select([
    F.count(
        F.when(F.col(c).isNaN(), c)
    ).alias(c) for c in str_col
]).show()

25/10/21 13:21:13 WARN TaskSetManager: Stage 7 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+--------+---------+----------+
|Quantity|UnitPrice|CustomerID|
+--------+---------+----------+
|       0|        0|    135080|
+--------+---------+----------+



There's 135,080 records with missing ID

In [10]:
total_count = df.count()
missing_count = df.where(F.col("CustomerID").isNaN()).count()

25/10/21 13:21:13 WARN TaskSetManager: Stage 10 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 13:21:13 WARN TaskSetManager: Stage 13 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


In [11]:
print(f"{round(missing_count/total_count*100, 2)}%")

24.93%


Remove from the dataset

In [12]:
df = df.na.drop(subset=['CustomerID'])
df.count()

25/10/21 13:21:13 WARN TaskSetManager: Stage 16 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


406829

## Customer ID

In [13]:
# Convert from double to string
df = df.withColumn('CustomerID', F.col("CustomerID").try_cast("int").try_cast("string"))
df.select("CustomerID").show(5)


+----------+
|CustomerID|
+----------+
|     17850|
|     17850|
|     17850|
|     17850|
|     17850|
+----------+
only showing top 5 rows


25/10/21 13:21:14 WARN TaskSetManager: Stage 19 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [14]:
df.schema['CustomerID'].dataType

StringType()

In [15]:
# Number of unique customers
df.select("CustomerID").distinct().count()

25/10/21 13:21:14 WARN TaskSetManager: Stage 20 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


4372

There are 4372 unique customers

## Purchase by country

In [16]:
# Purchases made by each country
df.groupBy("Country").agg(F.count_distinct("CustomerID").alias('country_count')).show() 

25/10/21 13:21:14 WARN TaskSetManager: Stage 26 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+------------------+-------------+
|           Country|country_count|
+------------------+-------------+
|            Sweden|            8|
|         Singapore|            1|
|           Germany|           95|
|               RSA|            1|
|            France|           87|
|            Greece|            4|
|European Community|            1|
|           Belgium|           25|
|           Finland|           12|
|             Malta|            2|
|       Unspecified|            4|
|             Italy|           15|
|              EIRE|            3|
|         Lithuania|            1|
|            Norway|           10|
|             Spain|           31|
|           Denmark|            9|
|            Israel|            4|
|           Iceland|            1|
|   Channel Islands|            9|
+------------------+-------------+
only showing top 20 rows


In [17]:
# Order the above table 
df.groupBy('Country').agg(F.count_distinct('CustomerID').alias('country_count')).orderBy(F.desc('country_count')).show()

25/10/21 13:21:15 WARN TaskSetManager: Stage 32 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------------+-------------+
|        Country|country_count|
+---------------+-------------+
| United Kingdom|         3950|
|        Germany|           95|
|         France|           87|
|          Spain|           31|
|        Belgium|           25|
|    Switzerland|           21|
|       Portugal|           19|
|          Italy|           15|
|        Finland|           12|
|        Austria|           11|
|         Norway|           10|
|        Denmark|            9|
|Channel Islands|            9|
|      Australia|            9|
|    Netherlands|            9|
|         Sweden|            8|
|         Cyprus|            8|
|          Japan|            8|
|         Poland|            6|
|         Greece|            4|
+---------------+-------------+
only showing top 20 rows


## Dates

When was the most recent purchase made by a customer on the e-commerse platform?

To find when the latest purchase was made on the platform, we need to convert the 'InvoiceDate' column into a timestamp format and use the 'max()' function in PySpark.

In [18]:
df = df.withColumn(
    "date" # New column name 
    , F.coalesce( # Coalesce function returns the first none-null value (evaluated left to right)
        F.to_timestamp(F.col("InvoiceDate"), "yy/MM/dd HH:mm"), # First attempt: tries to parse InvoiceDate using the pattern yy/MM/dd HH:mm and returns a Timestamp (e.g. 2017-07-15 14:30:00)
        F.to_timestamp(F.col("InvoiceDate"), "yyyy-MM-dd HH:mm:ss"),# If the first attempt fails and returns null, then it tries the next pattern
        F.to_timestamp(F.col("InvoiceDate")) # Best effort fallback 
    )
)
df.show(5)

+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|
|   536365|   84029G|KNITTED UNION FLA...|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|
|   536365|   84029E|RED WOOLLY HOTTIE...|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingd

25/10/21 13:21:16 WARN TaskSetManager: Stage 38 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [20]:
df.select(F.max("date")).show()

25/10/21 13:23:02 WARN TaskSetManager: Stage 39 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+-------------------+
|          max(date)|
+-------------------+
|2011-12-09 12:50:00|
+-------------------+



When was the earliest purchase made by a customer on the e-commerse platform?

In [21]:
df.select(F.min("date")).show()

25/10/21 13:23:06 WARN TaskSetManager: Stage 42 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+-------------------+
|          min(date)|
+-------------------+
|2010-12-01 08:26:00|
+-------------------+



# Data Pre-Processing

**RFM** is commonly used in marketing to evaluate a client's value based on their:

1. **Recency**: How recently has each customer made a purchase?
2. **Frequency**: How often have they bought something?
3. **Monetary Value**: How much money do they spend on average when making purchases?


## Recency

First, let's calculate the value of recency - the latest date and time a purchase was made on the platform. 

### Get the latest date

**There are three ways to do this: (adding a column with calculated constant)**

1. .collect() + .lit()
2. Using a cross join with aggregate
3. Using a window function

---

**Comparison:**

---
**1. .collect() + lit()**
- Not distributed (brings data to driver)
- Simple
- Best for small datasets

**2. crossJoin**
- Distributed
- Clear
- Best for medium to large datasets

**3. window function**
- Distributed
- Elegant, no join
- Best for when using window operations already

#### 1. .collect() + .lit()

**F.lit()** (short for literal) wraps a Python value (number, string, data, etc) into a **column expression** that Spark can handle.

In PySpark, all expressions inside withColumn() must be **Spark column objects**, not plain Python values. 

This is because Spark doesn't know how to distribute a pure Python literal (e.g. such as integer of 123) across the cluster automatically in that context.

This is why you need functions.lit() (short for literal) to wrap a Python value (number, string, date, etc) into a **column expression** that Spark can handle. 

In [ ]:
# Get the recent date
max_date = df.agg(F.max("InvoiceDate").alias("max_date")).collect()[0]["max_date"]
df = df.withColumn("from_date", F.lit(max_date))
df.show()


25/10/21 12:32:55 WARN TaskSetManager: Stage 31 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|   84029G|KNITTED UNION FLA...|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-

25/10/21 12:32:55 WARN TaskSetManager: Stage 34 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


**In this case:**
- max_date is a **single Python value** (after .collect() brings it from Spark to the driver)
- To broadcast it as a constant column to all rows, you must turn it into a **Spark column**
- That's exactly what F.lit() does

#### 2. Using a cross join with an aggregate

In [ ]:
# Aggregate to find the max InvoiceDate (still a DataFrame, not collected)
max_df = df.agg(F.max("InvoiceDate").alias("from_date2")) # Produces a 1-row DataFrame 
max_df.show()

# Cross join to add it to all rows
df = df.crossJoin(max_df) # Attaches the single row to every row of df (since max_df has only one row, the join is efficient)

df.show()

25/10/21 12:32:56 WARN TaskSetManager: Stage 35 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+-------------------+
|         from_date2|
+-------------------+
|2011-12-09 12:50:00|
+-------------------+



25/10/21 12:32:56 WARN TaskSetManager: Stage 38 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:56 WARN TaskSetManager: Stage 41 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+-------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|         from_date2|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+-------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|2011-12-09 12:50:00|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|2011-12-09 12:50:00|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|2011-12-09 12

#### 3. Using a window function

In [ ]:
from pyspark.sql import Window

df = df.withColumn(
    "from_date3",
    F.max("InvoiceDate").over(Window.rowsBetween(Window.unboundedPreceding, Window.unboundedFollowing)) # defines a windor over the entire dataset (from first to last row)
)

df.show()

25/10/21 12:32:56 WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
25/10/21 12:32:56 WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
25/10/21 12:32:56 WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
25/10/21 12:32:56 WARN TaskSetManager: Stage 42 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:56 WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
25/10/21 12:32:56 WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
25/10/21 12:32:56 WARN WindowExec

+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+-------------------+-------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|         from_date2|         from_date3|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+-------------------+-------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|2011-12-09 12:50:00|2011-12-09 12:50:00|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|2011-12-09 12:50:00|2011-12-09 12:50:00|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-0

In [ ]:
# Drop the last two columns
df = df.drop("from_date2", "from_date3")
df.show()

+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|
|   536365|   84029G|KNITTED UNION FLA...|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-

25/10/21 12:32:57 WARN TaskSetManager: Stage 48 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:57 WARN TaskSetManager: Stage 51 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


### Assign a receny score to each customer

We will subtract every date in the dataframe from the earliest date. This will tell us how recently a customer was seen in the dataframe. A value 0 indicates the lowest recency, as it will be assigned to the person who was seen making a purchase on the earliest date.

In [ ]:
# Calculate the the number of days between the max date and the purchase date
# df2 = df.withColumn("recency", col("date").cast("long") - col("from_date").cast("long")) # Convert date to long (integer)

df.select(F.col('date').cast("long")).show(5)
df.select(F.col('from_date').cast("long")).show(5)

df_rec = df.withColumn("recency", F.col("date").cast("long") - F.col("from_date").cast("long")) # Convert date to long (integer)
df_rec.show(5)

25/10/21 12:32:57 WARN TaskSetManager: Stage 52 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:57 WARN TaskSetManager: Stage 55 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


+----------+
|      date|
+----------+
|1291191960|
|1291191960|
|1291191960|
|1291191960|
|1291191960|
+----------+
only showing top 5 rows


25/10/21 12:32:57 WARN TaskSetManager: Stage 56 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:58 WARN TaskSetManager: Stage 59 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


+----------+
| from_date|
+----------+
|1323435000|
|1323435000|
|1323435000|
|1323435000|
|1323435000|
+----------+
only showing top 5 rows


25/10/21 12:32:58 WARN TaskSetManager: Stage 60 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+---------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|  recency|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+---------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|-32243040|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|-32243040|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|-32243040|
|   536365|   84029G|KNITTED UNION FLA...|       6|2

25/10/21 12:32:58 WARN TaskSetManager: Stage 63 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [ ]:
# Create a window function to add row number 
w = Window.partitionBy("CustomerID").orderBy(F.desc("recency"))
df_rec.withColumn("rn", F.row_number().over(w)).show()

25/10/21 12:32:58 WARN TaskSetManager: Stage 64 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:58 WARN TaskSetManager: Stage 67 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+---------+---------+--------------------+--------+-------------------+---------+----------+-------+-------------------+-------------------+--------+---+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|Country|               date|          from_date| recency| rn|
+---------+---------+--------------------+--------+-------------------+---------+----------+-------+-------------------+-------------------+--------+---+
|   581180|    23497|CLASSIC CHROME BI...|      12|2011-12-07 15:52:00|     1.45|     12347|Iceland|2011-12-07 15:52:00|2011-12-09 12:50:00| -161880|  1|
|   581180|    23552|BICYCLE PUNCTURE ...|       6|2011-12-07 15:52:00|     2.08|     12347|Iceland|2011-12-07 15:52:00|2011-12-09 12:50:00| -161880|  2|
|   581180|    21064|BOOM BOX SPEAKER ...|      24|2011-12-07 15:52:00|     1.25|     12347|Iceland|2011-12-07 15:52:00|2011-12-09 12:50:00| -161880|  3|
|   581180|   84625A|PINK NEW BAROQUEC...|      24|2011-12-07 15:52:00|     

### Select the most recent purchase

One customer can make multiple purchases at different times. We need to select only the last time they were seen buying a product (the latest), as this is indicative of when the most recent purchase was made:

In [ ]:
# Filter by the latest purchase
df_rec = df_rec.withColumn("rn", F.row_number().over(w)).filter(F.col("rn")==1).drop("rn")

In [ ]:
df_rec = df_rec.select(['CustomerID', 'recency'])
df_rec.show()

25/10/21 12:32:59 WARN TaskSetManager: Stage 70 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:32:59 WARN TaskSetManager: Stage 73 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+----------+---------+
|CustomerID|  recency|
+----------+---------+
|      NULL|    -8640|
|     12346|-28089180|
|     12347|  -161880|
|     12348| -6482220|
|     12349| -1565940|
|     12350|-26772540|
|     12352| -3103980|
|     12353|-17611380|
|     12354|-20047140|
|     12355|-18489660|
|     12356| -1915800|
|     12357| -2839380|
|     12358|   -95040|
|     12359|  -610140|
|     12360| -4487280|
|     12361|-24793140|
|     12362|  -249000|
|     12363| -9430320|
|     12364|  -613680|
|     12365|-25137960|
+----------+---------+
only showing top 20 rows


In [ ]:
df_rec.select('CustomerID').distinct().count()

25/10/21 12:34:07 WARN TaskSetManager: Stage 102 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:34:07 WARN TaskSetManager: Stage 105 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


4373

## Frequency

In [ ]:
# Calculate the number of purchases each customer made
df_freq = df.groupBy('CustomerID').agg(F.count('InvoiceDate').alias('frequency'))

df_freq.show()

25/10/21 12:33:01 WARN TaskSetManager: Stage 86 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:33:01 WARN TaskSetManager: Stage 89 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+----------+---------+
|CustomerID|frequency|
+----------+---------+
|     16250|       24|
|     15574|      168|
|     15555|      925|
|     15271|      275|
|     17757|      742|
|     17551|       43|
|     13187|       37|
|     16549|      981|
|     14525|      298|
|     18283|      756|
|     13107|       60|
|     16303|      167|
|     13174|      314|
|     13027|       26|
|     12957|      244|
|     17128|       14|
|     14810|      255|
|     16654|       31|
|     17491|      110|
|     17917|       94|
+----------+---------+
only showing top 20 rows


## Monetary Value

Total amount spent by each customer in the DataFrame

In [ ]:
# UnitPrice * Quantity
m_val = df.withColumn(
    "TotalAmount",
    F.col("Quantity").cast("double") * F.col("UnitPrice").cast("double")
)
m_val.show()

+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+------------------+
|InvoiceNo|StockCode|         Description|Quantity|        InvoiceDate|UnitPrice|CustomerID|       Country|               date|          from_date|       TotalAmount|
+---------+---------+--------------------+--------+-------------------+---------+----------+--------------+-------------------+-------------------+------------------+
|   536365|   85123A|WHITE HANGING HEA...|       6|2010-12-01 08:26:00|     2.55|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|15.299999999999999|
|   536365|    71053| WHITE METAL LANTERN|       6|2010-12-01 08:26:00|     3.39|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|             20.34|
|   536365|   84406B|CREAM CUPID HEART...|       8|2010-12-01 08:26:00|     2.75|     17850|United Kingdom|2010-12-01 08:26:00|2011-12-09 12:50:00|              22.0

25/10/21 12:33:02 WARN TaskSetManager: Stage 92 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:33:02 WARN TaskSetManager: Stage 95 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
Exception ignored in: <_io.BufferedWriter name=5>
Traceback (most recent call last):
  File "/Users/aiomote/Data Science Projects/datacamp_spark_tutorial/.venv/lib/python3.13/site-packages/pyspark/python/lib/pyspark.zip/pyspark/daemon.py", line 200, in manager
BrokenPipeError: [Errno 32] Broken pipe


In [ ]:
# Find the total amount spent by each customer:
m_val = m_val.groupBy('CustomerID').agg(sum('TotalAmount').alias('monetary_value'))
m_val.show()


25/10/21 12:33:02 WARN TaskSetManager: Stage 96 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:33:02 WARN TaskSetManager: Stage 99 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+----------+------------------+
|CustomerID|    monetary_value|
+----------+------------------+
|     16250|            389.44|
|     15574| 702.2500000000002|
|     15555| 4758.200000000001|
|     15271|2485.8199999999997|
|     17757| 5585.489999999999|
|     17551|            306.84|
|     13187|236.01999999999995|
|     16549| 4154.640000000001|
|     14525| 4221.370000000001|
|     18283| 2094.879999999999|
|     13107|1524.0700000000004|
|     16303| 5305.830000000002|
|     13174|           2817.03|
|     13027|            6912.0|
|     12957|3979.0400000000004|
|     17128|            157.09|
|     14810|2087.8800000000006|
|     16654|            594.97|
|     17491|           3541.92|
|     17917|            341.28|
+----------+------------------+
only showing top 20 rows


## Combine All

In [ ]:
final_df = df_rec.join(df_freq,
                       on = 'CustomerID',
                       how = 'inner')

final_df.show(5)

25/10/21 12:35:22 WARN TaskSetManager: Stage 111 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:22 WARN TaskSetManager: Stage 114 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:22 WARN TaskSetManager: Stage 115 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:23 WARN TaskSetManager: Stage 123 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:23 WARN TaskSetManager: Stage 126 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:23 WARN TaskSetManager: Stage 127 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


+----------+---------+---------+
|CustomerID|  recency|frequency|
+----------+---------+---------+
|     16250|-22542180|       24|
|     15574|-15299340|      168|
|     15555| -1024260|      925|
|     15271|  -599940|      275|
|     17757|   -76740|      742|
+----------+---------+---------+
only showing top 5 rows


In [ ]:
final_df.select("CustomerID").distinct().count()

25/10/21 12:35:43 WARN TaskSetManager: Stage 132 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:43 WARN TaskSetManager: Stage 135 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.
25/10/21 12:35:43 WARN TaskSetManager: Stage 136 contains a task of very large size (3433 KiB). The maximum recommended task size is 1000 KiB.


4372